### PR TITLE
Fixed source generator endpoint delegate handling for nullable scenarios

### LINKED ISSUE #54172: Request delegate generator produces CS8073 for ValueTask<T?> results
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

For delegates that return `ValueTask<T?>? the compiler emits
```
Warning	CS8073	The result of the expression is always 'false' since a value of type 'ValueTask<Person>' is never equal to 'null' of type 'ValueTask<Person>?'
```

This is especially cumbersome when 
```xml
<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
is set.

### Expected Behavior

No warning is produced.

### Steps To Reproduce

```c#
WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
WebApplication app = builder.Build();

Person[] persons = [new Person("Batman"), new Person("Me")];

// No problem
Person? SearchPerson0(string name) => Array.Find(persons, p => p.Name == name);

// Here the ValueTask causes the CS8073
ValueTask<Person?> SearchPerson1(string name)
{
    Person? person = Array.Find(persons, p => p.Name == name);
    return ValueTask.FromResult(person);
}

// This is fine, as Task is a reference type
Task<Person?> SearchPerson2(string name)
{
    Person? person = Array.Find(persons, p => p.Name == name);
    return Task.FromResult(person);
}

app.MapGet("person0/{name}", SearchPerson0);
app.MapGet("person1/{name}", SearchPerson1);
app.MapGet("person2/{name}", SearchPerson2);

app.Run();

public record Person(string Name);
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
    </PropertyGroup>

    <PropertyGroup>
        <IncludeOpenAPIAnalyzers>true</IncludeOpenAPIAnalyzers>

        <EnableRequestDelegateGenerator>true</EnableRequestDelegateGenerator>
        <InterceptorsPreviewNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.AspNetCore.Http.Generated</InterceptorsPreviewNamespaces>
    </PropertyGroup>

</Project>
```

### Exceptions (if any)

_No response_

### .NET Version

8.0.200

### Anything else?

If this got already fixed for .NET 9 is there a chance to backport it to .NET 8 as `<TreatWarningsAsErrors>` needs some workarounds to make it compile and .NET 8 is a LTS?

### PR DESCRIPTION
**Pull request by tcortega**
Fixed source generator endpoint delegate handling for nullable scenarios by storing the `WrappedResponseType` and using it to check for nullability.

Fixes #54172

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs [hunk 1]
```diff
@@ -427,23 +427,23 @@ public async Task MapAction_NoJsonTypeInfoResolver_ThrowsException()
...
+    public static IEnumerable<object[]> NullResultWithNullAnnotation
     {
         get
         {
             return new List<object[]>
             {
                 new object[] { "IResult? () => null", "The IResult returned by the Delegate must not be null." },
                 new object[] { "Task<IResult?>? () => null", "The Task returned by the Delegate must not be null." },
-                new object[] { "Task<bool?>? () => null", "The Task returned by the Delegate must not be null." },
+                new object[] { "Task<bool>? () => null", "The Task returned by the Delegate must not be null." },
```
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:438]: Can we retain the existing `Task<bool?>?` scenario here? I don't believe the change you made will break this and we want the right thing to happen regardless.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs [hunk 2]
```diff
@@ -458,6 +458,21 @@ public async Task RequestDelegateThrowsInvalidOperationExceptionOnNullDelegate(s
         Assert.Equal(message, exception.Message);
     }
 
+    [Theory]
+    [InlineData("Task<bool> () => null!")]
+    [InlineData("Task<bool?> () => null!")]
+    public async Task AwaitableRequestDelegateThrowsNullReferenceExceptionOnUnannotatedNullDelegate(string innerSource)
+    {
+        var source = $"""
```
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:466]:
```suggestion
        var source = $"""
app.MapGet("/", {innerSource});
""";
```

**Review by captainsafia — APPROVED** [review]

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs (hunk 1 above)
Comment (tcortega) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:438]: We sure can

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs (hunk 2 above)
Comment (tcortega) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:466]: @captainsafia Solved, but may I ask why?

**Pull request merged**